libfwupd/fwupd-client.c

```diff
@@ -491,6 +494,11 @@ fwupd_client_connect_get_bus_cb (GObject *source,
 		g_task_return_error (task, g_steal_pointer (&error));
 		return;
 	}
+
+	/* we want the GDBusProxy to emit signals in the thread default
+	 * main context of the thread which called fwupd_client_new()
+	 * not the first one that caused a fwupd_client_connect_async()... */
+	g_main_context_push_thread_default (priv->main_ctx_prxy);
```

Comment:
I think the solution for the Qt problem is to actually drop this push/pop pair, and instead change the `fwupd_client_properties_changed_cb()` and `fwupd_client_signal_cb()` functions to `g_idle_add_full()` the signal emissions through to the `main_ctx_prxy`. That has the same effect as these lines, but would avoid the critical warning you’re seeing from this push line.

The critical warning happens because this `GMainContext` is being pushed in the worker thread, while also being iterated in the main thread.

Reply:
> functions to g_idle_add_full() the signal emissions through to the main_ctx_prxy

Doesn't that imply g_idle_source_new(), g_source_set_callback() then g_source_attach() rather than g_idle_add_full()? I don't think you can specify the ctx for g_idle_add_full()

Reply:
Yes, you’re right.

Reply:
Something like 50f149648a04072b5887de7011b3189fd8760d5b perhaps? I think it's sane to share the priv->idle_id for signals and properties, right? I'm just trying to avoid queuing up hundreds of events if there is nothing servicing the main context.